cli: Fix JSON output for BPF conntrack & NAT tables dump

qmonnet opened the pull request:
JSON output is broken for conntrack and NAT tables dump with the cli (See #2976). Let's fix it. Dump arrays containing key/value tuples, so we can let the JSON parser decompose the entries into atomic values.

pchaigno commented on cilium/cmd/bpf_ct_list.go:
```diff
@@ -69,9 +76,15 @@ func dumpCt(eID string) {
-			if err := command.PrintOutput(m); err != nil {
-				os.Exit(1)
+			callback := func(key bpf.MapKey, value bpf.MapValue) {
+				keyval := ctKeyVal{MapKey: key.(ctmap.CtKey).ToHost(), MapValue: *value.(*ctmap.CtEntry)}
```
Why is the `.ToHost()` required? Was it broken before?

qmonnet replied:
It is not _required_ strictly speaking, but without it the `SourcePort` and `DestPort` from the key will be printed as they are stored in the BPF map, i.e. network byte order. It's not “broken” but not really convenient either, because you have to post-process the output to get the correct values for the ports.

pchaigno replied:
Hm, okay. Looks like there's the same issue with the `RevNat` field only in `CtEntry`.

qmonnet replied:
Ah, I missed it. There is no `ToHost()` available for the `CtEntry`, I'll see if I can add something.

Also looking into unit tests, but it requires mock maps or something.

joestringer replied:
I'd be interested to hear what kind of mock map infrastructure you'd like. I've thought given the various maps we have, it'd be nice to have a generic infrastructure across all maps for testing purposes, eventually based upon github.com/cilium/ebpf . But I haven't looked at the problem recently enough to share any particularly useful insights.

joestringer replied:
oh, and I see that the `test/runtime/cli.go` does have some runtime tests for the CLI, there may be an option there.

pchaigno reviewed (commented):
Looks good! I have one question below. Also wondering if we should add unit tests for this? Could be as simple as checking the output if correctly formatted JSON (e.g., `| python -m json.tool`).

pchaigno commented on cilium/cmd/bpf_ct_list_test.go:
```diff
@@ -0,0 +1,170 @@
+
+var _ = Suite(&BPFCtListSuite{})
+
+var (
```
Aren't most of these constants?

qmonnet replied:
In theory yes, but the different initialisers used to set the values return non-const types, so I get e.g. (from `go vet`):

```
cilium/cmd/bpf_ct_list_test.go:43:13:
        (types.IPv4 literal) (value of type types.IPv4) is not constant
```

If I try to replace `var` with `const` here.

pchaigno commented on cilium/cmd/bpf_ct_list_test.go:
```diff
@@ -0,0 +1,170 @@
+
+	ctMap0 := ctmap.NewCtMockMap()
+	ctMap0.InsertRecord(&ctKey4, ctValue)
+	ctMap0.InsertRecord(&ctKey4, ctValue)
```
Couldn't you pass these records directly as an array of `ctMockMapRecord` to `NewCtMockMap()`? Then you can get rid of `InsertRecord()` and initialize `ctMaps` as an array in a single call.

qmonnet replied:
I came up with:
``` go
	ctMaps := [2]ctmap.CtMap{
		ctmap.NewCtMockMap(
			[]ctmap.CtMockMapRecord{
				ctmap.CtMockMapRecord{
					Key:   &ctKey4,
					Value: &ctValue,
				},
				ctmap.CtMockMapRecord{
					Key:   &ctKey4,
					Value: &ctValue,
				},
			},
		),
		ctmap.NewCtMockMap(
			[]ctmap.CtMockMapRecord{
				ctmap.CtMockMapRecord{
					Key:   &ctKey4,
					Value: &ctValue,
				},
			},
		),
	}
```
Is this what you had in mind?

pchaigno replied:
:+1:

pchaigno commented on cilium/cmd/bpf_ct_list_test.go:
```diff
@@ -0,0 +1,170 @@
+	ctMap1 := ctmap.NewCtMockMap()
+	ctMap1.InsertRecord(&ctKey4, ctValue)
+
+	ctMaps := make([]ctmap.CtMap, 2)
```
This slice has a fixed size so could be declared as an array.

pchaigno reviewed (commented):
That last commit looks good to me. Only have a few nits below.

pchaigno commented on cilium/cmd/bpf_ct_list_test.go:
```diff
@@ -0,0 +1,194 @@
+	c.Assert(ctDump[0].MapKey.DestPort, Equals, uint16(80))
+	c.Assert(ctDump[0].MapKey.NextHeader, Equals, u8proto.U8proto(6))
+	c.Assert(ctDump[0].MapValue.RevNAT, Equals, uint16(27))
+	c.Assert(ctDump[0].MapValue.LastRxReport, Equals, uint32(7777))
```
Any chance you could turn these lines into a `checker.DeepEquals` `Check()`?

qmonnet replied:
Probably, what's the difference exactly?

joestringer replied:
`checker.DeepEquals` is a pretty-printer that provides diff-style comparison of observed vs. expected output. It makes interpreting the tests much easier by printing the failure inline, rather than just getting something like "Assert failed on line 42".

pchaigno replied:
Plus, compared to simple `Equals`, you would only need a single line to check the whole object.

qmonnet replied:
Ah, so I can compare directly the whole entry? Sounds good, thanks! Will do. And thanks for the explanations.

qmonnet replied:
So that would be:
```
    c.Assert(ctDump[0].MapKey, checker.DeepEquals,
        ctMaps[0].(*ctmap.CtMockMap).Entries[0].Key.(*ctmap.CtKey4).TupleKey4)
```

But the values aren't actually equal:
```
 tuple.TupleKey4{
     DestAddr:   {0xa, 0xa, 0xa, 0x1},
     SourceAddr: {0xa, 0xa, 0xa, 0x2},
-    DestPort:   0x50,
-    SourcePort: 0x350b,
+    DestPort:   0x5000,
+    SourcePort: 0xb35,
     NextHeader: 0x6,
     Flags:      0x7b,
 }
```

Because I chose to `ntohs()` the fields that were in network byte order for readability. Which makes me wonder if this is a good choice. Do we want them readable in the JSON, at the expense of JSON entry not being a 100% exact mirror of the internal struct, or do we want parity? (If the JSON is processed at some point and we try to map it to internal structs, parity is probably better? And surely for readability there is plain output anyway?)

pchaigno replied:
I'd go for 100% exact mirror. We could also override the `Unmarshal()` function, but it's really not worth it here...

pchaigno commented on pkg/maps/ctmap/ctmap_mock.go:
```diff
@@ -0,0 +1,70 @@
+
+// CtMockMap implements the CtMap interface and can be used for unit tests.
+type CtMockMap struct {
+	Entries []CtMockMapRecord
```
Could you use `[]ctKeyVal` here? That would remove some code and simplify the `DeepEquals` `Assert()`.

qmonnet replied:
I can and it will remove one struct indeed, but the `Assert()` won't be any simpler. This is because we unmarshall JSON into `[]ctRecord4` and `[]ctRecord6`, which don't have a `ctmap.CtKey` as key (or unmarshalling would not work), so the `DeepEquals` find different objects.

qmonnet committed:
cli: fix JSON output for BPF conntrack table dump

JSON output for BPF conntrack table dump is broken: We pass map objects
to the JSON printer, which prints information on the maps but does not
iterate on (or print) its entries. The final output may not even a real
JSON object if we call the printer on several maps (several JSON objects
are dumped successively, not gathered in an array).

Let's fix JSON output. Collect all the entries (key/values) from the
different maps, and pass them to the json package so it can print
something nice. The result now looks like this:

    # cilium bpf ct global list -o json
    [
      {
        "Key": {
          "DestAddr": [
            192,
            168,
            33,
            11
          ],
          "SourceAddr": [
            10,
            16,
            196,
            61
          ],
          "DestPort": 40624,
          "SourcePort": 6443,
          "NextHeader": 6,
          "Flags": 1
        },
        "Value": {
          "RxPackets": 1,
          "RxBytes": 706,
          "TxPackets": 1,
          "TxBytes": 54,
          "Lifetime": 49472,
          "Flags": 18,
          "RevNAT": 0,
          "TxFlagsSeen": 4,
          "RxFlagsSeen": 2,
          "SourceSecurityID": 1,
          "LastTxReport": 27872,
          "LastRxReport": 27872
        }
      },

      ...

    ]

Fixes: #2976
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
cli: fix JSON output for BPF NAT table dump

Fix JSON output for dumping the NAT tables with the cli, in a way
similar to what we did for the conntrack tables.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
ctmap: add mock maps and unit test for BPF conntrack JSON dump

Add mock maps to the ctmap package. They can be used for unit tests.

Speaking of: Add a unit test for JSON dump of the conntrack maps via the
cli. We dump and parse JSON, then we check that some of the values are
consistent.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
nat: add mock maps and tests for NAT map JSON dump

Similarly to ctmap, add mock maps and unit tests for JSON dump for NAT
maps. The code is very similar to the equivalent for ctmap. Make generic
and reuse dumpAndRead() from bpf_ct_list_test.go.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno commented on cilium/cmd/bpf_ct_list_test.go:
```diff
@@ -0,0 +1,204 @@
+	c.Assert(ctDump[0].Key, checker.DeepEquals,
+		ctMaps[0].(*ctmap.CtMockMap).Entries[0].Key.(*ctmap.CtKey4).TupleKey4)
+	c.Assert(ctDump[0].Value, checker.DeepEquals,
+		ctMaps[0].(*ctmap.CtMockMap).Entries[0].Value)
```
Can be simplified a little (but not sure if worth it):
```suggestion
	ctRecordDump := ctmap.CtMapRecord{&ctmap.CtKey4{ctDump[0].Key}, ctDump[0].Value}
	c.Assert(ctRecordDump, checker.DeepEquals, ctMaps[0].(*ctmap.CtMockMap).Entries[0])
```

aanm reviewed (approved):
@pchaigno @joestringer the coverage decreased because @qmonnet added a test for `cilium/cmd` package which was not being accounted in coveralls because it does not have any test. The coverage seems to be accounted per package that contains a `_test.go`

joestringer commented on pkg/maps/nat/nat_mock.go:
```diff
@@ -0,0 +1,63 @@
+
+// NatMockMap implements the NatMap interface and can be used for unit tests.
+type NatMockMap struct {
+	Entries []NatMapRecord
```
This is nice to see as an example of the kind of mocking we would want to be able to properly unit-test the core of the daemon. I wonder if you have thought about how we might centralize these mocks so that we have a single implementation you can use to instantiate a memory-backed "BPF" map in a generic way without having to duplicate the similar base functionality in each package?

FWIW we have a `pkg/testutils` that hosts a bunch of things like this.

Separately, we've also been pondering about whether it makes sense to have such a mock framework as part of the `cilium/ebpf` repository. Without examples like these, we haven't really known what such mocks would need to look like just yet.

qmonnet replied:
I did think about centralising them, but I wasn't sure there was enough interest at this time and preferred not to wander too far in that direction, apart from reusing the `dumpAndRead()` bit. But yeah, definitely something we could consider for the future. Most of the code for NAT and CT mock maps is very close, it should not be too hard to figure out something. Depending on what other map types or use cases you have in mind, it could add some challenge, of course.

joestringer replied:
One mock to rule them all! ...well yeah I guess just mainly the ones we're interacting with from Cilium userspace which are primarily hash-or-similar: LRU, LPM, hashmap. I don't think we even at the moment need to care *too* much about the fact that lookups for LPM work differently, as BPF LPM map lookups are only done from the (Envoy) proxy and not from cilium-agent [citation needed].

joestringer reviewed (approved):
Minor new discussion point, more future-looking and not relevant to the acceptance of this PR.
